### PR TITLE
feat(router): add support for request/response transformers

### PR DESCRIPTION
**Pull request by jonball4**
[feat(router): add support for request/response transformers](https://github.com/BitGo/api-ts/pull/680/commits/3a54035a1a0430422d3dbaadff0484f87398f54f)

### CODE REVIEW

**Review by stevenharrisBG — CHANGES_REQUESTED** [review]
Please add jsdoc and update README
It really isn't at all clear how to use these changes without deep diving.

File: packages/typed-express-router/src/types.ts [hunk 1]
```diff
@@ -37,25 +54,125 @@ export type AfterEncodedResponseSentFn = (
...
 };
 
+export type WrappedRouteOptions = UncheckedWrappedRouteOptions &
+  TransformedRouteOptions & { onDecodeError?: OnDecodeErrorFn };
+
 export type WrappedRouterOptions = express.RouterOptions & WrappedRouteOptions;
 
+type TypeTransformer<D, U> = (decoded: D) => U;
+
+type ApiMethodRequest<
```
Comment (manav-c) [inline packages/typed-express-router/src/types.ts:105]:
```suggestion
type ApiRequest<
```

File: packages/typed-express-router/src/types.ts [hunk 2]
```diff
@@ -37,25 +54,125 @@ export type AfterEncodedResponseSentFn = (
...
+  Spec extends ApiSpec,
+  ApiName extends keyof Spec,
+  Method extends keyof Spec[ApiName] & HttpMethod,
+> = NonNullable<Spec[ApiName][Method]>['request'];
+
+type ApiMethodRequestType<
```
Comment (manav-c) [inline packages/typed-express-router/src/types.ts:111]:
```suggestion
type ApiRequestType<
```

File: packages/typed-express-router/src/errors.ts [hunk 3]
```diff
@@ -20,3 +20,21 @@ export function defaultOnEncodeError(
   res.status(500).end();
   console.warn(`Error in route handler: ${err}`);
 }
+
+export function defaultOnRequestTransformError(
+  err: unknown,
+  _req: express.Request,
+  res: express.Response,
+) {
+  res.status(500).end();
```
[the remark this answers is not in the document]
Comment (manav-c) [reply packages/typed-express-router/src/errors.ts:29]: should update the default response error handler as well

**Comment by manav-c** [comment]
Discussed with Jon on Slack - summary of my thoughts on these changes: These changes would make it clear to devs that there is a pattern of transforms that stems from thinking about the distinction between layers, like data and service -- but the transformation functions can easily be abused (just applying transformations in the route handler to meet the expectations of the transformation) or ignored completely. While these changes prompt devs think about transforms, I feel like the distinction pattern is more of a mindset that we need to cultivate within the team and am not sure if this PR helps with that specifically.

**Pull request closed without merging**